{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 47,
   "metadata": {},
   "outputs": [],
   "source": [
    "import numpy as np\n",
    "import pandas as pd\n",
    "import xml.etree.ElementTree as ET\n",
    "import xmlschema\n",
    "import networkx as nx\n",
    "from xml.etree import cElementTree as ElementTree\n",
    "import matplotlib.pyplot as plt\n",
    "import re\n",
    "import csv\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 48,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/Users/janiceyang/anaconda3/lib/python3.7/site-packages/ipykernel_launcher.py:35: FutureWarning: The behavior of this method will change in future versions.  Use specific 'len(elem)' or 'elem is not None' test instead.\n",
      "/Users/janiceyang/anaconda3/lib/python3.7/site-packages/ipykernel_launcher.py:11: FutureWarning: The behavior of this method will change in future versions.  Use specific 'len(elem)' or 'elem is not None' test instead.\n",
      "  # This is added back by InteractiveShellApp.init_path()\n"
     ]
    }
   ],
   "source": [
    "\"\"\"\n",
    "DISEASE-gene from Orphanet: get only link for VWD for now? (VWD<-VWF)\n",
    "\"\"\"\n",
    "\n",
    "filepath = \"../Data/Orphanet_disease-gene.xml\"\n",
    "\n",
    "# REFERENCE 1\n",
    "class XmlListConfig(list):\n",
    "    def __init__(self, aList):\n",
    "        for element in aList:\n",
    "            if element:\n",
    "                # treat like dict\n",
    "                if len(element) == 1 or element[0].tag != element[1].tag:\n",
    "                    self.append(XmlDictConfig(element))\n",
    "                # treat like list\n",
    "                elif element[0].tag == element[1].tag:\n",
    "                    self.append(XmlListConfig(element))\n",
    "            elif element.text:\n",
    "                text = element.text.strip()\n",
    "                if text:\n",
    "                    self.append(text)\n",
    "\n",
    "class XmlDictConfig(dict):\n",
    "    '''\n",
    "    Example usage:\n",
    "\n",
    "    >>> tree = ElementTree.parse('your_file.xml')\n",
    "    >>> root = tree.getroot()\n",
    "    >>> xmldict = XmlDictConfig(root)\n",
    "    '''\n",
    "    def __init__(self, parent_element):\n",
    "        if parent_element.items():\n",
    "            self.update(dict(parent_element.items()))\n",
    "        for element in parent_element:\n",
    "            if element:\n",
    "                # treat like dict - we assume that if the first two tags\n",
    "                # in a series are different, then they are all different.\n",
    "                if len(element) == 1 or element[0].tag != element[1].tag:\n",
    "                    aDict = XmlDictConfig(element)\n",
    "                # treat like list - we assume that if the first two tags\n",
    "                # in a series are the same, then the rest are the same.\n",
    "                else:\n",
    "                    # here, we put the list in dictionary; the key is the\n",
    "                    # tag name the list elements all share in common, and\n",
    "                    # the value is the list itself \n",
    "                    aDict = {element[0].tag: XmlListConfig(element)}\n",
    "                # if the tag has attributes, add those to the dict\n",
    "                if element.items():\n",
    "                    aDict.update(dict(element.items()))\n",
    "                self.update({element.tag: aDict})\n",
    "            # this assumes that if you've got an attribute in a tag,\n",
    "            # you won't be having any text. This may or may not be a \n",
    "            # good idea -- time will tell. It works for the way we are\n",
    "            # currently doing XML configuration files...\n",
    "#             elif element.items():\n",
    "#                 self.update({element.tag: dict(element.items())})\n",
    "            # finally, if there are no child tags and no attributes, extract\n",
    "            # the text\n",
    "            else:\n",
    "                self.update({element.tag: element.text})\n",
    "                \n",
    "\n",
    "etree = ET.parse(filepath) #create an ElementTree object \n",
    "root = etree.getroot()\n",
    "xmldict = XmlDictConfig(root)\n",
    "\n",
    "disease_list = xmldict['DisorderList']['Disorder']\n",
    "\n",
    "## Make a dataframe with only relevant fields: \n",
    "    # OrphaNumber (Orphanet ID)\n",
    "    # Name (disease name)\n",
    "    # DisorderGeneAssociationList['Count'] (number of genes associated with the disease)\n",
    "    ## May need loop for:\n",
    "    # DisorderGeneAssociationList['DisorderGeneAssociation']['Gene']['Name'] (name of associated gene)\n",
    "    # DisorderGeneAssociationList['DisorderGeneAssociation']['Gene']['ExternalReferenceList']['ExternalReference'] (Gene IDs and sources)\n",
    "    # DisorderGeneAssociationList['DisorderGeneAssociation']['Gene']['DisorderGeneAssociationType']['Name']?\n",
    "    # DisorderGeneAssociationList['DisorderGeneAssociation']['Gene']['DisorderGeneAssociationStatus']['Name'] (whether association is 'Assessed')\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 49,
   "metadata": {
    "collapsed": true
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "{'id': '17601',\n",
       " 'OrphaNumber': '166024',\n",
       " 'Name': 'Multiple epiphyseal dysplasia, Al-Gazali type',\n",
       " 'DisorderGeneAssociationList': {'count': '1',\n",
       "  'DisorderGeneAssociation': {'SourceOfValidation': '22587682[PMID]',\n",
       "   'Gene': {'id': '20160',\n",
       "    'OrphaNumber': '268061',\n",
       "    'Name': 'kinesin family member 7',\n",
       "    'Symbol': 'KIF7',\n",
       "    'SynonymList': {'count': '1', 'Synonym': 'JBTS12'},\n",
       "    'ExternalReferenceList': {'ExternalReference': [{'id': '57240',\n",
       "       'Source': 'Ensembl',\n",
       "       'Reference': 'ENSG00000166813'},\n",
       "      {'id': '51758', 'Source': 'Genatlas', 'Reference': 'KIF7'},\n",
       "      {'id': '51756', 'Source': 'HGNC', 'Reference': '30497'},\n",
       "      {'id': '51757', 'Source': 'OMIM', 'Reference': '611254'},\n",
       "      {'id': '97306', 'Source': 'Reactome', 'Reference': 'Q2M1P5'},\n",
       "      {'id': '51759', 'Source': 'SwissProt', 'Reference': 'Q2M1P5'}],\n",
       "     'count': '6'},\n",
       "    'LocusList': {'count': '1', 'Locus': '\\n              '}},\n",
       "   'DisorderGeneAssociationType': {'id': '17949',\n",
       "    'Name': 'Disease-causing germline mutation(s) in'},\n",
       "   'DisorderGeneAssociationStatus': {'id': '17991', 'Name': 'Assessed'}}}}"
      ]
     },
     "execution_count": 49,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "disease_list[0]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 50,
   "metadata": {
    "collapsed": true
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Number of terms:3802\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>OrphaNumber</th>\n",
       "      <th>Disease_name</th>\n",
       "      <th>GeneAssociationCount</th>\n",
       "      <th>Associated_genes</th>\n",
       "      <th>Gene_IDs</th>\n",
       "      <th>Association_types</th>\n",
       "      <th>Association_statuses</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>166024</td>\n",
       "      <td>Multiple epiphyseal dysplasia, Al-Gazali type</td>\n",
       "      <td>1</td>\n",
       "      <td>[kinesin family member 7]</td>\n",
       "      <td>[{'ExternalReference': [{'id': '57240', 'Sourc...</td>\n",
       "      <td>[Disease-causing germline mutation(s) in]</td>\n",
       "      <td>[Assessed]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>93</td>\n",
       "      <td>Aspartylglucosaminuria</td>\n",
       "      <td>1</td>\n",
       "      <td>[aspartylglucosaminidase]</td>\n",
       "      <td>[{'ExternalReference': [{'id': '126323', 'Sour...</td>\n",
       "      <td>[Disease-causing germline mutation(s) in]</td>\n",
       "      <td>[Assessed]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>166035</td>\n",
       "      <td>Brachydactyly-short stature-retinitis pigmento...</td>\n",
       "      <td>1</td>\n",
       "      <td>[CWC27 spliceosome associated protein homolog]</td>\n",
       "      <td>[\\n            ]</td>\n",
       "      <td>[Disease-causing germline mutation(s) in]</td>\n",
       "      <td>[Assessed]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>585</td>\n",
       "      <td>Multiple sulfatase deficiency</td>\n",
       "      <td>1</td>\n",
       "      <td>[sulfatase modifying factor 1]</td>\n",
       "      <td>[{'ExternalReference': [{'id': '56683', 'Sourc...</td>\n",
       "      <td>[Disease-causing germline mutation(s) in]</td>\n",
       "      <td>[Assessed]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>118</td>\n",
       "      <td>Beta-mannosidosis</td>\n",
       "      <td>1</td>\n",
       "      <td>[mannosidase beta]</td>\n",
       "      <td>[{'ExternalReference': [{'id': '100307', 'Sour...</td>\n",
       "      <td>[Disease-causing germline mutation(s) in]</td>\n",
       "      <td>[Assessed]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>166068</td>\n",
       "      <td>Pontocerebellar hypoplasia type 5</td>\n",
       "      <td>1</td>\n",
       "      <td>[tRNA splicing endonuclease subunit 54]</td>\n",
       "      <td>[{'ExternalReference': [{'id': '58222', 'Sourc...</td>\n",
       "      <td>[Disease-causing germline mutation(s) in]</td>\n",
       "      <td>[Assessed]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>166063</td>\n",
       "      <td>Pontocerebellar hypoplasia type 4</td>\n",
       "      <td>1</td>\n",
       "      <td>[tRNA splicing endonuclease subunit 54]</td>\n",
       "      <td>[{'ExternalReference': [{'id': '58222', 'Sourc...</td>\n",
       "      <td>[Disease-causing germline mutation(s) in]</td>\n",
       "      <td>[Assessed]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>166078</td>\n",
       "      <td>Von Willebrand disease type 1</td>\n",
       "      <td>1</td>\n",
       "      <td>[von Willebrand factor]</td>\n",
       "      <td>[{'ExternalReference': [{'id': '60135', 'Sourc...</td>\n",
       "      <td>[Disease-causing germline mutation(s) in]</td>\n",
       "      <td>[Assessed]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>166073</td>\n",
       "      <td>Pontocerebellar hypoplasia type 6</td>\n",
       "      <td>1</td>\n",
       "      <td>[arginyl-tRNA synthetase 2, mitochondrial]</td>\n",
       "      <td>[{'ExternalReference': [{'id': '60133', 'Sourc...</td>\n",
       "      <td>[Disease-causing germline mutation(s) in]</td>\n",
       "      <td>[Assessed]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>166084</td>\n",
       "      <td>Von Willebrand disease type 2A</td>\n",
       "      <td>1</td>\n",
       "      <td>[von Willebrand factor]</td>\n",
       "      <td>[{'ExternalReference': [{'id': '60135', 'Sourc...</td>\n",
       "      <td>[Disease-causing germline mutation(s) in]</td>\n",
       "      <td>[Assessed]</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "   OrphaNumber                                       Disease_name  \\\n",
       "0       166024      Multiple epiphyseal dysplasia, Al-Gazali type   \n",
       "1           93                             Aspartylglucosaminuria   \n",
       "2       166035  Brachydactyly-short stature-retinitis pigmento...   \n",
       "3          585                      Multiple sulfatase deficiency   \n",
       "4          118                                  Beta-mannosidosis   \n",
       "5       166068                  Pontocerebellar hypoplasia type 5   \n",
       "6       166063                  Pontocerebellar hypoplasia type 4   \n",
       "7       166078                      Von Willebrand disease type 1   \n",
       "8       166073                  Pontocerebellar hypoplasia type 6   \n",
       "9       166084                     Von Willebrand disease type 2A   \n",
       "\n",
       "   GeneAssociationCount                                Associated_genes  \\\n",
       "0                     1                       [kinesin family member 7]   \n",
       "1                     1                       [aspartylglucosaminidase]   \n",
       "2                     1  [CWC27 spliceosome associated protein homolog]   \n",
       "3                     1                  [sulfatase modifying factor 1]   \n",
       "4                     1                              [mannosidase beta]   \n",
       "5                     1         [tRNA splicing endonuclease subunit 54]   \n",
       "6                     1         [tRNA splicing endonuclease subunit 54]   \n",
       "7                     1                         [von Willebrand factor]   \n",
       "8                     1      [arginyl-tRNA synthetase 2, mitochondrial]   \n",
       "9                     1                         [von Willebrand factor]   \n",
       "\n",
       "                                            Gene_IDs  \\\n",
       "0  [{'ExternalReference': [{'id': '57240', 'Sourc...   \n",
       "1  [{'ExternalReference': [{'id': '126323', 'Sour...   \n",
       "2                                   [\\n            ]   \n",
       "3  [{'ExternalReference': [{'id': '56683', 'Sourc...   \n",
       "4  [{'ExternalReference': [{'id': '100307', 'Sour...   \n",
       "5  [{'ExternalReference': [{'id': '58222', 'Sourc...   \n",
       "6  [{'ExternalReference': [{'id': '58222', 'Sourc...   \n",
       "7  [{'ExternalReference': [{'id': '60135', 'Sourc...   \n",
       "8  [{'ExternalReference': [{'id': '60133', 'Sourc...   \n",
       "9  [{'ExternalReference': [{'id': '60135', 'Sourc...   \n",
       "\n",
       "                           Association_types Association_statuses  \n",
       "0  [Disease-causing germline mutation(s) in]           [Assessed]  \n",
       "1  [Disease-causing germline mutation(s) in]           [Assessed]  \n",
       "2  [Disease-causing germline mutation(s) in]           [Assessed]  \n",
       "3  [Disease-causing germline mutation(s) in]           [Assessed]  \n",
       "4  [Disease-causing germline mutation(s) in]           [Assessed]  \n",
       "5  [Disease-causing germline mutation(s) in]           [Assessed]  \n",
       "6  [Disease-causing germline mutation(s) in]           [Assessed]  \n",
       "7  [Disease-causing germline mutation(s) in]           [Assessed]  \n",
       "8  [Disease-causing germline mutation(s) in]           [Assessed]  \n",
       "9  [Disease-causing germline mutation(s) in]           [Assessed]  "
      ]
     },
     "execution_count": 50,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "OrphaNumbers = []\n",
    "Disease_names = []\n",
    "GeneAssociationCounts = []\n",
    "Associated_genes = []\n",
    "Gene_IDs = []\n",
    "Association_types = []\n",
    "Association_statuses = []\n",
    "counter = -1\n",
    "\n",
    "for disease in disease_list:\n",
    "    OrphaNumber = int(disease['OrphaNumber'])\n",
    "    dnames = disease['Name']\n",
    "    Count = int(disease['DisorderGeneAssociationList']['count'])\n",
    "    \n",
    "    associations = disease['DisorderGeneAssociationList']['DisorderGeneAssociation'] # dict if 1 association, list if >1\n",
    "    genes = []\n",
    "    gene_symbols = []\n",
    "    geneIDs = []\n",
    "    types = []\n",
    "    statuses = []\n",
    "    if Count > 1:\n",
    "        for i, association in enumerate(associations):\n",
    "            genes.append(associations[i]['Gene']['Name'])\n",
    "            gene_symbols.append(associations[i]['Gene']['Symbol'])\n",
    "            geneIDs.append(associations[i]['Gene']['ExternalReferenceList'])\n",
    "            types.append(associations[i]['DisorderGeneAssociationType']['Name'])\n",
    "            statuses.append(associations[i]['DisorderGeneAssociationStatus']['Name'])\n",
    "    else:\n",
    "        genes.append(associations['Gene']['Name'])\n",
    "        gene_symbols.append(associations['Gene']['Symbol'])\n",
    "        geneIDs.append(associations['Gene']['ExternalReferenceList'])\n",
    "        types.append(associations['DisorderGeneAssociationType']['Name'])\n",
    "        statuses.append(associations['DisorderGeneAssociationStatus']['Name'])        \n",
    "    \n",
    "    # Update column lists\n",
    "    OrphaNumbers.append(OrphaNumber)\n",
    "    Disease_names.append(dnames)\n",
    "    GeneAssociationCounts.append(Count)\n",
    "    Associated_genes.append(genes)\n",
    "    Gene_IDs.append(geneIDs)\n",
    "    Association_types.append(types)\n",
    "    Association_statuses.append(statuses)\n",
    "    \n",
    "    counter += 1\n",
    "\n",
    "print(\"Number of terms:\" + str(counter))\n",
    "# colnames = ['OrphaNumber', 'Disease_name', 'GeneAssociationCount', \n",
    "#             'Associated_genes', 'Gene_IDs', 'Association_types', 'Association_statuses'] \n",
    "coldict = {'OrphaNumber': OrphaNumbers,\n",
    "          'Disease_name': Disease_names,\n",
    "          'GeneAssociationCount': GeneAssociationCounts,\n",
    "          'Associated_genes': Associated_genes,\n",
    "          'Gene_IDs': Gene_IDs,\n",
    "          'Association_types': Association_types,\n",
    "          'Association_statuses': Association_statuses}\n",
    "\n",
    "disease_df = pd.DataFrame(coldict)\n",
    "disease_df.to_csv(\"../Data/Orphanet_disease-gene_df.csv\")\n",
    "disease_df.head(10)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 51,
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "3415\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>OrphaNumber</th>\n",
       "      <th>Disease_name</th>\n",
       "      <th>GeneAssociationCount</th>\n",
       "      <th>Associated_genes</th>\n",
       "      <th>Gene_IDs</th>\n",
       "      <th>Association_types</th>\n",
       "      <th>Association_statuses</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>166024</td>\n",
       "      <td>Multiple epiphyseal dysplasia, Al-Gazali type</td>\n",
       "      <td>1</td>\n",
       "      <td>[kinesin family member 7]</td>\n",
       "      <td>[{'ExternalReference': [{'id': '57240', 'Sourc...</td>\n",
       "      <td>[Disease-causing germline mutation(s) in]</td>\n",
       "      <td>[Assessed]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>93</td>\n",
       "      <td>Aspartylglucosaminuria</td>\n",
       "      <td>1</td>\n",
       "      <td>[aspartylglucosaminidase]</td>\n",
       "      <td>[{'ExternalReference': [{'id': '126323', 'Sour...</td>\n",
       "      <td>[Disease-causing germline mutation(s) in]</td>\n",
       "      <td>[Assessed]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>166035</td>\n",
       "      <td>Brachydactyly-short stature-retinitis pigmento...</td>\n",
       "      <td>1</td>\n",
       "      <td>[CWC27 spliceosome associated protein homolog]</td>\n",
       "      <td>[\\n            ]</td>\n",
       "      <td>[Disease-causing germline mutation(s) in]</td>\n",
       "      <td>[Assessed]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>585</td>\n",
       "      <td>Multiple sulfatase deficiency</td>\n",
       "      <td>1</td>\n",
       "      <td>[sulfatase modifying factor 1]</td>\n",
       "      <td>[{'ExternalReference': [{'id': '56683', 'Sourc...</td>\n",
       "      <td>[Disease-causing germline mutation(s) in]</td>\n",
       "      <td>[Assessed]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>118</td>\n",
       "      <td>Beta-mannosidosis</td>\n",
       "      <td>1</td>\n",
       "      <td>[mannosidase beta]</td>\n",
       "      <td>[{'ExternalReference': [{'id': '100307', 'Sour...</td>\n",
       "      <td>[Disease-causing germline mutation(s) in]</td>\n",
       "      <td>[Assessed]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>166068</td>\n",
       "      <td>Pontocerebellar hypoplasia type 5</td>\n",
       "      <td>1</td>\n",
       "      <td>[tRNA splicing endonuclease subunit 54]</td>\n",
       "      <td>[{'ExternalReference': [{'id': '58222', 'Sourc...</td>\n",
       "      <td>[Disease-causing germline mutation(s) in]</td>\n",
       "      <td>[Assessed]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>166063</td>\n",
       "      <td>Pontocerebellar hypoplasia type 4</td>\n",
       "      <td>1</td>\n",
       "      <td>[tRNA splicing endonuclease subunit 54]</td>\n",
       "      <td>[{'ExternalReference': [{'id': '58222', 'Sourc...</td>\n",
       "      <td>[Disease-causing germline mutation(s) in]</td>\n",
       "      <td>[Assessed]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>166078</td>\n",
       "      <td>Von Willebrand disease type 1</td>\n",
       "      <td>1</td>\n",
       "      <td>[von Willebrand factor]</td>\n",
       "      <td>[{'ExternalReference': [{'id': '60135', 'Sourc...</td>\n",
       "      <td>[Disease-causing germline mutation(s) in]</td>\n",
       "      <td>[Assessed]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>166073</td>\n",
       "      <td>Pontocerebellar hypoplasia type 6</td>\n",
       "      <td>1</td>\n",
       "      <td>[arginyl-tRNA synthetase 2, mitochondrial]</td>\n",
       "      <td>[{'ExternalReference': [{'id': '60133', 'Sourc...</td>\n",
       "      <td>[Disease-causing germline mutation(s) in]</td>\n",
       "      <td>[Assessed]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>166084</td>\n",
       "      <td>Von Willebrand disease type 2A</td>\n",
       "      <td>1</td>\n",
       "      <td>[von Willebrand factor]</td>\n",
       "      <td>[{'ExternalReference': [{'id': '60135', 'Sourc...</td>\n",
       "      <td>[Disease-causing germline mutation(s) in]</td>\n",
       "      <td>[Assessed]</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "   OrphaNumber                                       Disease_name  \\\n",
       "0       166024      Multiple epiphyseal dysplasia, Al-Gazali type   \n",
       "1           93                             Aspartylglucosaminuria   \n",
       "2       166035  Brachydactyly-short stature-retinitis pigmento...   \n",
       "3          585                      Multiple sulfatase deficiency   \n",
       "4          118                                  Beta-mannosidosis   \n",
       "5       166068                  Pontocerebellar hypoplasia type 5   \n",
       "6       166063                  Pontocerebellar hypoplasia type 4   \n",
       "7       166078                      Von Willebrand disease type 1   \n",
       "8       166073                  Pontocerebellar hypoplasia type 6   \n",
       "9       166084                     Von Willebrand disease type 2A   \n",
       "\n",
       "   GeneAssociationCount                                Associated_genes  \\\n",
       "0                     1                       [kinesin family member 7]   \n",
       "1                     1                       [aspartylglucosaminidase]   \n",
       "2                     1  [CWC27 spliceosome associated protein homolog]   \n",
       "3                     1                  [sulfatase modifying factor 1]   \n",
       "4                     1                              [mannosidase beta]   \n",
       "5                     1         [tRNA splicing endonuclease subunit 54]   \n",
       "6                     1         [tRNA splicing endonuclease subunit 54]   \n",
       "7                     1                         [von Willebrand factor]   \n",
       "8                     1      [arginyl-tRNA synthetase 2, mitochondrial]   \n",
       "9                     1                         [von Willebrand factor]   \n",
       "\n",
       "                                            Gene_IDs  \\\n",
       "0  [{'ExternalReference': [{'id': '57240', 'Sourc...   \n",
       "1  [{'ExternalReference': [{'id': '126323', 'Sour...   \n",
       "2                                   [\\n            ]   \n",
       "3  [{'ExternalReference': [{'id': '56683', 'Sourc...   \n",
       "4  [{'ExternalReference': [{'id': '100307', 'Sour...   \n",
       "5  [{'ExternalReference': [{'id': '58222', 'Sourc...   \n",
       "6  [{'ExternalReference': [{'id': '58222', 'Sourc...   \n",
       "7  [{'ExternalReference': [{'id': '60135', 'Sourc...   \n",
       "8  [{'ExternalReference': [{'id': '60133', 'Sourc...   \n",
       "9  [{'ExternalReference': [{'id': '60135', 'Sourc...   \n",
       "\n",
       "                           Association_types Association_statuses  \n",
       "0  [Disease-causing germline mutation(s) in]           [Assessed]  \n",
       "1  [Disease-causing germline mutation(s) in]           [Assessed]  \n",
       "2  [Disease-causing germline mutation(s) in]           [Assessed]  \n",
       "3  [Disease-causing germline mutation(s) in]           [Assessed]  \n",
       "4  [Disease-causing germline mutation(s) in]           [Assessed]  \n",
       "5  [Disease-causing germline mutation(s) in]           [Assessed]  \n",
       "6  [Disease-causing germline mutation(s) in]           [Assessed]  \n",
       "7  [Disease-causing germline mutation(s) in]           [Assessed]  \n",
       "8  [Disease-causing germline mutation(s) in]           [Assessed]  \n",
       "9  [Disease-causing germline mutation(s) in]           [Assessed]  "
      ]
     },
     "execution_count": 51,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Disease associations to use: 'Disease-causing' substring\n",
    "# Other potential associations: 'Part of fusion gene', 'Major susceptibility factor', 'Role in phenotype' substrings\n",
    "\n",
    "# Keep only entries with causal disease associations:\n",
    "for i, association in enumerate(disease_df['Association_types']):\n",
    "    if 'disease-causing' not in str(association).lower():\n",
    "        disease_df = disease_df.drop(i)\n",
    "        \n",
    "print(len(disease_df))\n",
    "disease_df.head(10)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "[{'id': '57240', 'Source': 'Ensembl', 'Reference': 'ENSG00000166813'},\n",
       " {'id': '51758', 'Source': 'Genatlas', 'Reference': 'KIF7'},\n",
       " {'id': '51756', 'Source': 'HGNC', 'Reference': '30497'},\n",
       " {'id': '51757', 'Source': 'OMIM', 'Reference': '611254'},\n",
       " {'id': '97306', 'Source': 'Reactome', 'Reference': 'Q2M1P5'},\n",
       " {'id': '51759', 'Source': 'SwissProt', 'Reference': 'Q2M1P5'}]"
      ]
     },
     "execution_count": 21,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "for ref in disease_df.iloc[0]['Gene_IDs'][0]['ExternalReference']:\n",
    "    if ref['Source'] == 'HGNC':\n",
    "        snode = ref['Reference']\n",
    "# disease_df.iloc[0]['Gene_IDs'][0]['ExternalReference']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 25,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "True"
      ]
     },
     "execution_count": 25,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "dict == type(disease_df.iloc[0]['Gene_IDs'][0])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 52,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>source_node</th>\n",
       "      <th>snode_type</th>\n",
       "      <th>gene_name</th>\n",
       "      <th>target_node</th>\n",
       "      <th>tnode_type</th>\n",
       "      <th>disease_name</th>\n",
       "      <th>assoc_type</th>\n",
       "      <th>assoc_status</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>30497</td>\n",
       "      <td>gene</td>\n",
       "      <td>kinesin family member 7</td>\n",
       "      <td>166024</td>\n",
       "      <td>disease</td>\n",
       "      <td>Multiple epiphyseal dysplasia, Al-Gazali type</td>\n",
       "      <td>Disease-causing germline mutation(s) in</td>\n",
       "      <td>Assessed</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>318</td>\n",
       "      <td>gene</td>\n",
       "      <td>aspartylglucosaminidase</td>\n",
       "      <td>93</td>\n",
       "      <td>disease</td>\n",
       "      <td>Aspartylglucosaminuria</td>\n",
       "      <td>Disease-causing germline mutation(s) in</td>\n",
       "      <td>Assessed</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>20376</td>\n",
       "      <td>gene</td>\n",
       "      <td>sulfatase modifying factor 1</td>\n",
       "      <td>585</td>\n",
       "      <td>disease</td>\n",
       "      <td>Multiple sulfatase deficiency</td>\n",
       "      <td>Disease-causing germline mutation(s) in</td>\n",
       "      <td>Assessed</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>6831</td>\n",
       "      <td>gene</td>\n",
       "      <td>mannosidase beta</td>\n",
       "      <td>118</td>\n",
       "      <td>disease</td>\n",
       "      <td>Beta-mannosidosis</td>\n",
       "      <td>Disease-causing germline mutation(s) in</td>\n",
       "      <td>Assessed</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>27561</td>\n",
       "      <td>gene</td>\n",
       "      <td>tRNA splicing endonuclease subunit 54</td>\n",
       "      <td>166068</td>\n",
       "      <td>disease</td>\n",
       "      <td>Pontocerebellar hypoplasia type 5</td>\n",
       "      <td>Disease-causing germline mutation(s) in</td>\n",
       "      <td>Assessed</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>27561</td>\n",
       "      <td>gene</td>\n",
       "      <td>tRNA splicing endonuclease subunit 54</td>\n",
       "      <td>166063</td>\n",
       "      <td>disease</td>\n",
       "      <td>Pontocerebellar hypoplasia type 4</td>\n",
       "      <td>Disease-causing germline mutation(s) in</td>\n",
       "      <td>Assessed</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>12726</td>\n",
       "      <td>gene</td>\n",
       "      <td>von Willebrand factor</td>\n",
       "      <td>166078</td>\n",
       "      <td>disease</td>\n",
       "      <td>Von Willebrand disease type 1</td>\n",
       "      <td>Disease-causing germline mutation(s) in</td>\n",
       "      <td>Assessed</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>21406</td>\n",
       "      <td>gene</td>\n",
       "      <td>arginyl-tRNA synthetase 2, mitochondrial</td>\n",
       "      <td>166073</td>\n",
       "      <td>disease</td>\n",
       "      <td>Pontocerebellar hypoplasia type 6</td>\n",
       "      <td>Disease-causing germline mutation(s) in</td>\n",
       "      <td>Assessed</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>12726</td>\n",
       "      <td>gene</td>\n",
       "      <td>von Willebrand factor</td>\n",
       "      <td>166084</td>\n",
       "      <td>disease</td>\n",
       "      <td>Von Willebrand disease type 2A</td>\n",
       "      <td>Disease-causing germline mutation(s) in</td>\n",
       "      <td>Assessed</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>735</td>\n",
       "      <td>gene</td>\n",
       "      <td>N-acylsphingosine amidohydrolase 1</td>\n",
       "      <td>333</td>\n",
       "      <td>disease</td>\n",
       "      <td>Farber disease</td>\n",
       "      <td>Disease-causing germline mutation(s) in</td>\n",
       "      <td>Assessed</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "  source_node snode_type                                 gene_name  \\\n",
       "0       30497       gene                   kinesin family member 7   \n",
       "1         318       gene                   aspartylglucosaminidase   \n",
       "2       20376       gene              sulfatase modifying factor 1   \n",
       "3        6831       gene                          mannosidase beta   \n",
       "4       27561       gene     tRNA splicing endonuclease subunit 54   \n",
       "5       27561       gene     tRNA splicing endonuclease subunit 54   \n",
       "6       12726       gene                     von Willebrand factor   \n",
       "7       21406       gene  arginyl-tRNA synthetase 2, mitochondrial   \n",
       "8       12726       gene                     von Willebrand factor   \n",
       "9         735       gene        N-acylsphingosine amidohydrolase 1   \n",
       "\n",
       "  target_node tnode_type                                   disease_name  \\\n",
       "0      166024    disease  Multiple epiphyseal dysplasia, Al-Gazali type   \n",
       "1          93    disease                         Aspartylglucosaminuria   \n",
       "2         585    disease                  Multiple sulfatase deficiency   \n",
       "3         118    disease                              Beta-mannosidosis   \n",
       "4      166068    disease              Pontocerebellar hypoplasia type 5   \n",
       "5      166063    disease              Pontocerebellar hypoplasia type 4   \n",
       "6      166078    disease                  Von Willebrand disease type 1   \n",
       "7      166073    disease              Pontocerebellar hypoplasia type 6   \n",
       "8      166084    disease                 Von Willebrand disease type 2A   \n",
       "9         333    disease                                 Farber disease   \n",
       "\n",
       "                                assoc_type assoc_status  \n",
       "0  Disease-causing germline mutation(s) in     Assessed  \n",
       "1  Disease-causing germline mutation(s) in     Assessed  \n",
       "2  Disease-causing germline mutation(s) in     Assessed  \n",
       "3  Disease-causing germline mutation(s) in     Assessed  \n",
       "4  Disease-causing germline mutation(s) in     Assessed  \n",
       "5  Disease-causing germline mutation(s) in     Assessed  \n",
       "6  Disease-causing germline mutation(s) in     Assessed  \n",
       "7  Disease-causing germline mutation(s) in     Assessed  \n",
       "8  Disease-causing germline mutation(s) in     Assessed  \n",
       "9  Disease-causing germline mutation(s) in     Assessed  "
      ]
     },
     "execution_count": 52,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Build graph! Dataframe columns: source node, source node attributes, target node, target node attributes, edge attributes\n",
    "NE_cols = ['source_node', 'snode_type', 'gene_name', \n",
    "           'target_node', 'tnode_type', 'disease_name', \n",
    "           'assoc_type', 'assoc_status']\n",
    "NE_df = pd.DataFrame(columns = NE_cols)\n",
    "\n",
    "for i, genelist in enumerate(disease_df['Associated_genes']):\n",
    "    for j, gene in enumerate(genelist):\n",
    "        if type(disease_df.iloc[i]['Gene_IDs'][0]) != dict:\n",
    "            pass\n",
    "        else:\n",
    "            for ref in disease_df.iloc[i]['Gene_IDs'][0]['ExternalReference']:\n",
    "                if ref['Source'] == 'HGNC': # Using HGNC ID since pretty much all databases reference it for genes\n",
    "                    snode = ref['Reference']\n",
    "                    snode_type = 'gene'\n",
    "                    gene_name = gene\n",
    "\n",
    "                    tnode = disease_df.iloc[i]['OrphaNumber']\n",
    "                    tnode_type = 'disease'\n",
    "                    disease_name = disease_df.iloc[i]['Disease_name']\n",
    "\n",
    "                    assoc_type = disease_df.iloc[i]['Association_types'][j]\n",
    "                    status = disease_df.iloc[i]['Association_statuses'][j]\n",
    "                    newrow = [snode, snode_type, gene_name, tnode, tnode_type, disease_name, assoc_type, status]\n",
    "                    NE_df.loc[len(NE_df)] = newrow\n",
    "                else:\n",
    "                    pass\n",
    "\n",
    "NE_df.head(10)\n",
    "                                                       "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 53,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Create diseases graph object\n",
    "disease_gene_graph = nx.from_pandas_edgelist(NE_df, source = 'source_node', target = 'target_node', \n",
    "                                             edge_attr = ['assoc_type', 'assoc_status'])\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 54,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Node attributes: key is node, value is dict of attribute-value pairs\n",
    "snode_attr = {node: {'node_type': NE_df.iloc[i]['snode_type'], \n",
    "                     'gene_name': NE_df.iloc[i]['gene_name']} for i, node in enumerate(NE_df['source_node'])}\n",
    "tnode_attr = {node: {'node_type': NE_df.iloc[i]['tnode_type'], \n",
    "                     'disease_name': NE_df.iloc[i]['disease_name']} for i, node in enumerate(NE_df['target_node'])}\n",
    "\n",
    "nx.set_node_attributes(disease_gene_graph, snode_attr)\n",
    "nx.set_node_attributes(disease_gene_graph, tnode_attr)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 55,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Create edgelist file? Test with BioNEV code\n",
    "filepath = '../Graphs/OrphaNet_disease_gene.edgelist'\n",
    "with open(filepath, 'wb') as file:\n",
    "    nx.write_edgelist(disease_gene_graph, file)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 56,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Visualize graph (not really helpful at this point)\n",
    "# nx.draw_networkx(disease_gene_graph, with_labels = False, nodelist = NE_df.source_node.tolist(), node_color = 'r', alpha = 0.5)\n",
    "# nx.draw_networkx(disease_gene_graph, with_labels = False, nodelist = NE_df.target_node.tolist(), node_color = 'b', alpha = 0.5)\n",
    "# plt.show()\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [],
   "source": [
    "\"\"\"\n",
    "DRUG-gene target from DrugBank*\n",
    "\n",
    "*Combine with KEGG? Or just use DB or KEGG alone\n",
    "\n",
    "\"\"\"\n",
    "filepath = \"../Data/DrugBank full database.xml\"\n",
    "xsdpath = \"../Data/drugbank_schema.xsd\"\n",
    "\n",
    "# XML schema object\n",
    "xs = xmlschema.XMLSchema(xsdpath)\n",
    "\n",
    "# Decode to dictionary\n",
    "### TAKES A WHILE (LIKE 30-ish minutes) TO RUN\n",
    "db_dict = xs.to_dict(filepath)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 57,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "dict_keys(['@xmlns', '@xmlns:xsi', '@xsi:schemaLocation', '@version', '@exported-on', 'drug'])"
      ]
     },
     "execution_count": 57,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "db_dict.keys() # drug data is stored in 'drug' key in list of dictionary form\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 160,
   "metadata": {
    "collapsed": true
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "{'@position': 1,\n",
       " 'id': 'BE0000565',\n",
       " 'name': 'Interleukin-1 receptor type 1',\n",
       " 'organism': 'Humans',\n",
       " 'actions': {'action': ['antagonist']},\n",
       " 'references': {'articles': {'article': [{'ref-id': 'A1826',\n",
       "     'pubmed-id': '17498496',\n",
       "     'citation': 'Tang YH, Zhang SP, Liang Y, Deng CQ: [Effects of Panax notoginseng saponins on mRNA expressions of interleukin-1 beta, its correlative factors and cysteinyl-aspartate specific protease after cerebral ischemia-reperfusion in rats]. Zhong Xi Yi Jie He Xue Bao. 2007 May;5(3):328-32.'},\n",
       "    {'ref-id': 'A1828',\n",
       "     'pubmed-id': '12817089',\n",
       "     'citation': 'Dayer JM: The pivotal role of interleukin-1 in the clinical manifestations of rheumatoid arthritis. Rheumatology (Oxford). 2003 May;42 Suppl 2:ii3-10.'},\n",
       "    {'ref-id': 'A1830',\n",
       "     'pubmed-id': '12355453',\n",
       "     'citation': 'Vamvakopoulos J, Green C, Metcalfe S: Genetic control of IL-1beta bioactivity through differential regulation of the IL-1 receptor antagonist. Eur J Immunol. 2002 Oct;32(10):2988-96.'},\n",
       "    {'ref-id': 'A1832',\n",
       "     'pubmed-id': '17083033',\n",
       "     'citation': 'Do H, Vasilescu A, Carpentier W, Meyer L, Diop G, Hirtzig T, Coulonges C, Labib T, Spadoni JL, Therwath A, Lathrop M, Matsuda F, Zagury JF: Exhaustive genotyping of the interleukin-1 family genes and associations with AIDS progression in a French cohort. J Infect Dis. 2006 Dec 1;194(11):1492-504. Epub 2006 Oct 26.'},\n",
       "    {'ref-id': 'A9',\n",
       "     'pubmed-id': '11752352',\n",
       "     'citation': 'Chen X, Ji ZL, Chen YZ: TTD: Therapeutic Target Database. Nucleic Acids Res. 2002 Jan 1;30(1):412-5.'},\n",
       "    {'ref-id': 'A1835',\n",
       "     'pubmed-id': '17352828',\n",
       "     'citation': 'So A, De Smedt T, Revaz S, Tschopp J: A pilot study of IL-1 inhibition by anakinra in acute gout. Arthritis Res Ther. 2007;9(2):R28.'},\n",
       "    {'ref-id': 'A1837',\n",
       "     'pubmed-id': '10210771',\n",
       "     'citation': 'Vannier E, Kaser A, Atkins MB, Fantuzzi G, Dinarello CA, Mier JW, Tilg H: Elevated circulating levels of soluble interleukin-1 receptor type II during interleukin-2 immunotherapy. Eur Cytokine Netw. 1999 Mar;10(1):37-42.'}]},\n",
       "  'textbooks': None,\n",
       "  'links': None,\n",
       "  'attachments': None},\n",
       " 'known-action': 'yes',\n",
       " 'polypeptide': [{'@id': 'P14778',\n",
       "   '@source': 'Swiss-Prot',\n",
       "   'name': 'Interleukin-1 receptor type 1',\n",
       "   'general-function': 'Transmembrane signaling receptor activity',\n",
       "   'specific-function': 'Receptor for IL1A, IL1B and IL1RN. After binding to interleukin-1 associates with the corecptor IL1RAP to form the high affinity interleukin-1 receptor complex which mediates interleukin-1-dependent activation of NF-kappa-B, MAPK and other pathways. Signaling involves the recruitment of adapter molecules such as TOLLIP, MYD88, and IRAK1 or IRAK2 via the respective TIR domains of the receptor/coreceptor subunits. Binds ligands with comparable affinity and binding of antagonist IL1RN prevents association with IL1RAP to form a signaling complex.',\n",
       "   'gene-name': 'IL1R1',\n",
       "   'locus': '2q12',\n",
       "   'cellular-location': 'Membrane',\n",
       "   'transmembrane-regions': '337-356',\n",
       "   'signal-regions': '1-17',\n",
       "   'theoretical-pi': '7.89',\n",
       "   'molecular-weight': '65401.91',\n",
       "   'chromosome-location': '2',\n",
       "   'organism': {'@ncbi-taxonomy-id': '9606', '$': 'Humans'},\n",
       "   'external-identifiers': {'external-identifier': [{'resource': 'HUGO Gene Nomenclature Committee (HGNC)',\n",
       "      'identifier': 'HGNC:5993'},\n",
       "     {'resource': 'GenAtlas', 'identifier': 'IL1R1'},\n",
       "     {'resource': 'GenBank Gene Database', 'identifier': 'X16896'},\n",
       "     {'resource': 'GenBank Protein Database', 'identifier': '33801'},\n",
       "     {'resource': 'UniProtKB', 'identifier': 'P14778'},\n",
       "     {'resource': 'UniProt Accession', 'identifier': 'IL1R1_HUMAN'}]},\n",
       "   'synonyms': {'synonym': ['CD121 antigen-like family member A',\n",
       "     'IL-1R-1',\n",
       "     'IL-1R-alpha',\n",
       "     'IL1R',\n",
       "     'IL1RA',\n",
       "     'IL1RT1',\n",
       "     'Interleukin-1 receptor alpha',\n",
       "     'Interleukin-1 receptor type I',\n",
       "     'p80']},\n",
       "   'amino-acid-sequence': {'@format': 'FASTA',\n",
       "    '$': '>lcl|BSEQ0018975|Interleukin-1 receptor type 1\\nMKVLLRLICFIALLISSLEADKCKEREEKIILVSSANEIDVRPCPLNPNEHKGTITWYKD\\nDSKTPVSTEQASRIHQHKEKLWFVPAKVEDSGHYYCVVRNSSYCLRIKISAKFVENEPNL\\nCYNAQAIFKQKLPVAGDGGLVCPYMEFFKNENNELPKLQWYKDCKPLLLDNIHFSGVKDR\\nLIVMNVAEKHRGNYTCHASYTYLGKQYPITRVIEFITLEENKPTRPVIVSPANETMEVDL\\nGSQIQLICNVTGQLSDIAYWKWNGSVIDEDDPVLGEDYYSVENPANKRRSTLITVLNISE\\nIESRFYKHPFTCFAKNTHGIDAAYIQLIYPVTNFQKHMIGICVTLTVIIVCSVFIYKIFK\\nIDIVLWYRDSCYDFLPIKASDGKTYDAYILYPKTVGEGSTSDCDIFVFKVLPEVLEKQCG\\nYKLFIYGRDDYVGEDIVEVINENVKKSRRLIIILVRETSGFSWLGGSSEEQIAMYNALVQ\\nDGIKVVLLELEKIQDYEKMPESIKFIKQKHGAIRWSGDFTQGPQSAKTRFWKNVRYHMPV\\nQRRSPSSKHQLLSPATKEKLQREAHVPLG'},\n",
       "   'gene-sequence': {'@format': 'FASTA',\n",
       "    '$': '>lcl|BSEQ0018976|Interleukin-1 receptor type 1 (IL1R1)\\nATGAAAGTGTTACTCAGACTTATTTGTTTCATAGCTCTACTGATTTCTTCTCTGGAGGCT\\nGATAAATGCAAGGAACGTGAAGAAAAAATAATTTTAGTGTCATCTGCAAATGAAATTGAT\\nGTTCGTCCCTGTCCTCTTAACCCAAATGAACACAAAGGCACTATAACTTGGTATAAAGAT\\nGACAGCAAGACACCTGTATCTACAGAACAAGCCTCCAGGATTCATCAACACAAAGAGAAA\\nCTTTGGTTTGTTCCTGCTAAGGTGGAGGATTCAGGACATTACTATTGCGTGGTAAGAAAT\\nTCATCTTACTGCCTCAGAATTAAAATAAGTGCAAAATTTGTGGAGAATGAGCCTAACTTA\\nTGTTATAATGCACAAGCCATATTTAAGCAGAAACTACCCGTTGCAGGAGACGGAGGACTT\\nGTGTGCCCTTATATGGAGTTTTTTAAAAATGAAAATAATGAGTTACCTAAATTACAGTGG\\nTATAAGGATTGCAAACCTCTACTTCTTGACAATATACACTTTAGTGGAGTCAAAGATAGG\\nCTCATCGTGATGAATGTGGCTGAAAAGCATAGAGGGAACTATACTTGTCATGCATCCTAC\\nACATACTTGGGCAAGCAATATCCTATTACCCGGGTAATAGAATTTATTACTCTAGAGGAA\\nAACAAACCCACAAGGCCTGTGATTGTGAGCCCAGCTAATGAGACAATGGAAGTAGACTTG\\nGGATCCCAGATACAATTGATCTGTAATGTCACCGGCCAGTTGAGTGACATTGCTTACTGG\\nAAGTGGAATGGGTCAGTAATTGATGAAGATGACCCAGTGCTAGGGGAAGACTATTACAGT\\nGTGGAAAATCCTGCAAACAAAAGAAGGAGTACCCTCATCACAGTGCTTAATATATCGGAA\\nATTGAAAGTAGATTTTATAAACATCCATTTACCTGTTTTGCCAAGAATACACATGGTATA\\nGATGCAGCATATATCCAGTTAATATATCCAGTCACTAATTTCCAGAAGCACATGATTGGT\\nATATGTGTCACGTTGACAGTCATAATTGTGTGTTCTGTTTTCATCTATAAAATCTTCAAG\\nATTGACATTGTGCTTTGGTACAGGGATTCCTGCTATGATTTTCTCCCAATAAAAGCTTCA\\nGATGGAAAGACCTATGACGCATATATACTGTATCCAAAGACTGTTGGGGAAGGGTCTACC\\nTCTGACTGTGATATTTTTGTGTTTAAAGTCTTGCCTGAGGTCTTGGAAAAACAGTGTGGA\\nTATAAGCTGTTCATTTATGGAAGGGATGACTACGTTGGGGAAGACATTGTTGAGGTCATT\\nAATGAAAACGTAAAGAAAAGCAGAAGACTGATTATCATTTTAGTCAGAGAAACATCAGGC\\nTTCAGCTGGCTGGGTGGTTCATCTGAAGAGCAAATAGCCATGTATAATGCTCTTGTTCAG\\nGATGGAATTAAAGTTGTCCTGCTTGAGCTGGAGAAAATCCAAGACTATGAGAAAATGCCA\\nGAATCGATTAAATTCATTAAGCAGAAACATGGGGCTATCCGCTGGTCAGGGGACTTTACA\\nCAGGGACCACAGTCTGCAAAGACAAGGTTCTGGAAGAATGTCAGGTACCACATGCCAGTC\\nCAGCGACGGTCACCTTCATCTAAACACCAGTTACTGTCACCAGCCACTAAGGAGAAACTG\\nCAAAGAGAGGCTCACGTGCCTCTCGGGTAG'},\n",
       "   'pfams': {'pfam': [{'identifier': 'PF07679', 'name': 'I-set'},\n",
       "     {'identifier': 'PF01582', 'name': 'TIR'},\n",
       "     {'identifier': 'PF13895', 'name': 'Ig_2'}]},\n",
       "   'go-classifiers': {'go-classifier': [{'category': 'component',\n",
       "      'description': 'axon'},\n",
       "     {'category': 'component', 'description': 'cell surface'},\n",
       "     {'category': 'component', 'description': 'extracellular space'},\n",
       "     {'category': 'component',\n",
       "      'description': 'integral component of plasma membrane'},\n",
       "     {'category': 'component', 'description': 'membrane'},\n",
       "     {'category': 'component', 'description': 'nucleus'},\n",
       "     {'category': 'component', 'description': 'plasma membrane'},\n",
       "     {'category': 'component', 'description': 'postsynaptic density'},\n",
       "     {'category': 'component', 'description': 'protein complex'},\n",
       "     {'category': 'function',\n",
       "      'description': 'interleukin-1 receptor activity'},\n",
       "     {'category': 'function',\n",
       "      'description': 'interleukin-1, Type I, activating receptor activity'},\n",
       "     {'category': 'function',\n",
       "      'description': 'platelet-derived growth factor receptor binding'},\n",
       "     {'category': 'function', 'description': 'signal transducer activity'},\n",
       "     {'category': 'function',\n",
       "      'description': 'transmembrane signaling receptor activity'},\n",
       "     {'category': 'process',\n",
       "      'description': 'cell surface receptor signaling pathway'},\n",
       "     {'category': 'process',\n",
       "      'description': 'cellular response to glucose stimulus'},\n",
       "     {'category': 'process', 'description': 'heat acclimation'},\n",
       "     {'category': 'process', 'description': 'immune response'},\n",
       "     {'category': 'process',\n",
       "      'description': 'interleukin-1-mediated signaling pathway'},\n",
       "     {'category': 'process', 'description': 'ovulation'},\n",
       "     {'category': 'process',\n",
       "      'description': 'regulation of inflammatory response'},\n",
       "     {'category': 'process',\n",
       "      'description': 'response to epidermal growth factor'},\n",
       "     {'category': 'process', 'description': 'response to interleukin-1'},\n",
       "     {'category': 'process', 'description': 'response to nitric oxide'},\n",
       "     {'category': 'process', 'description': 'response to radiation'},\n",
       "     {'category': 'process',\n",
       "      'description': 'response to transforming growth factor beta'}]}}]}"
      ]
     },
     "execution_count": 160,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Examine db_dict\n",
    "db_dict['drug'][24]['targets']['target'][0]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Number of drugs with polypeptide targets: 5232\n",
      "Number of drugs with non-polypeptide targets: 202\n",
      "Number of drugs in DrugBank total: 13339\n"
     ]
    }
   ],
   "source": [
    "# Create list of dictionaries, with each dict containing info for a single drug\n",
    "drugs_list = db_dict['drug']\n",
    "\n",
    "drugs_w_targets = list()\n",
    "counter_notpp = 0\n",
    "for drug in drugs_list:\n",
    "    if 'target' not in str(drug['targets']):\n",
    "        pass\n",
    "    elif 'polypeptide' not in str(drug['targets']['target'][0]):\n",
    "        counter_notpp += 1\n",
    "    elif 'HUGO Gene Nomenclature Committee (HGNC)' in str(drug['targets']['target'][0]['polypeptide'][0]['external-identifiers']['external-identifier']):\n",
    "        drugs_w_targets.append(drug)\n",
    "        \n",
    "print(\"Number of drugs with polypeptide targets: \" + str(len(drugs_w_targets)))\n",
    "print(\"Number of drugs with non-polypeptide targets: \" + str(counter_notpp))\n",
    "print(\"Number of drugs in DrugBank total: \" + str(len(drugs_list)))\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Need source(drug) columns: drugs_list<insert indices>\n",
    "    # drug_ID: id[i] if id[i]['@primary'] == True for id in ['drugbank-id']\n",
    "    # drug_name: ['name'] \n",
    "    # drug_action: ['targets']['target'][i]['actions']['action']\n",
    "    # drug_type: ['classification']['direct-parent']\n",
    "# Need target(gene) columns: drugs_list['targets']['target']<insert indices>\n",
    "    # target_ID: ['polypeptide'][i]['external-identifiers']['external-identifier'][i][]\n",
    "    # target_name: ['name']\n",
    "\n",
    "DB_cols = ['drug_ID', 'drug_name', 'drug_action', 'drug_type', 'snode_type',\n",
    "          'target_ID', 'target_name', 'tnode_type']\n",
    "DB_df = pd.DataFrame(columns = DB_cols)  \n",
    "# idx = 0\n",
    "for i, drug in enumerate(drugs_w_targets):\n",
    "    for ID in drug['drugbank-id']:\n",
    "        if ID['@primary'] == True:\n",
    "            drug_ID = ID['$']\n",
    "    drug_name = drug['name']\n",
    "    if 'classification' not in drug.keys():\n",
    "        drug_type = ''\n",
    "    else:\n",
    "        drug_type = drug['classification']['direct-parent']\n",
    "    snode_type = 'drug'\n",
    "    \n",
    "    for target in drug['targets']['target']:\n",
    "        if 'polypeptide' not in target.keys():\n",
    "            pass\n",
    "        elif 'HGNC' not in str(target['polypeptide'][0]['external-identifiers']['external-identifier'][0]['identifier']):\n",
    "            pass\n",
    "        else:\n",
    "            HGNC = target['polypeptide'][0]['external-identifiers']['external-identifier'][0]['identifier'].split(':')[1]\n",
    "            target_ID = HGNC\n",
    "            target_name = target['name']\n",
    "            tnode_type = 'target'\n",
    "            if target['actions'] == None:\n",
    "                drug_action = ''\n",
    "            else:\n",
    "                drug_action = target['actions']['action']\n",
    "            newrow = [drug_ID, drug_name, drug_action, drug_type, snode_type, target_ID, target_name, tnode_type]\n",
    "            DB_df.loc[len(DB_df)] = newrow\n",
    "#     print(idx)\n",
    "#     idx += 1\n",
    "    "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>drug_ID</th>\n",
       "      <th>drug_name</th>\n",
       "      <th>drug_action</th>\n",
       "      <th>drug_type</th>\n",
       "      <th>snode_type</th>\n",
       "      <th>target_ID</th>\n",
       "      <th>target_name</th>\n",
       "      <th>tnode_type</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>13414</th>\n",
       "      <td>DB14738</td>\n",
       "      <td>Turoctocog alfa pegol</td>\n",
       "      <td>[activator]</td>\n",
       "      <td>Peptides</td>\n",
       "      <td>drug</td>\n",
       "      <td>3551</td>\n",
       "      <td>Coagulation factor IX</td>\n",
       "      <td>target</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13415</th>\n",
       "      <td>DB14738</td>\n",
       "      <td>Turoctocog alfa pegol</td>\n",
       "      <td>[activator]</td>\n",
       "      <td>Peptides</td>\n",
       "      <td>drug</td>\n",
       "      <td>3528</td>\n",
       "      <td>Coagulation factor X</td>\n",
       "      <td>target</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13416</th>\n",
       "      <td>DB14738</td>\n",
       "      <td>Turoctocog alfa pegol</td>\n",
       "      <td>[binder]</td>\n",
       "      <td>Peptides</td>\n",
       "      <td>drug</td>\n",
       "      <td>3535</td>\n",
       "      <td>Prothrombin</td>\n",
       "      <td>target</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13417</th>\n",
       "      <td>DB14751</td>\n",
       "      <td>Mecasermin rinfabate</td>\n",
       "      <td>[agonist]</td>\n",
       "      <td>Peptides</td>\n",
       "      <td>drug</td>\n",
       "      <td>5465</td>\n",
       "      <td>Insulin-like growth factor 1 receptor</td>\n",
       "      <td>target</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13418</th>\n",
       "      <td>DB14751</td>\n",
       "      <td>Mecasermin rinfabate</td>\n",
       "      <td></td>\n",
       "      <td>Peptides</td>\n",
       "      <td>drug</td>\n",
       "      <td>6091</td>\n",
       "      <td>Insulin receptor</td>\n",
       "      <td>target</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13419</th>\n",
       "      <td>DB14751</td>\n",
       "      <td>Mecasermin rinfabate</td>\n",
       "      <td></td>\n",
       "      <td>Peptides</td>\n",
       "      <td>drug</td>\n",
       "      <td>5467</td>\n",
       "      <td>Cation-independent mannose-6-phosphate receptor</td>\n",
       "      <td>target</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13420</th>\n",
       "      <td>DB14754</td>\n",
       "      <td>Solriamfetol</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>drug</td>\n",
       "      <td>11049</td>\n",
       "      <td>Sodium-dependent dopamine transporter</td>\n",
       "      <td>target</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13421</th>\n",
       "      <td>DB14754</td>\n",
       "      <td>Solriamfetol</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>drug</td>\n",
       "      <td>11048</td>\n",
       "      <td>Sodium-dependent noradrenaline transporter</td>\n",
       "      <td>target</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13422</th>\n",
       "      <td>DB14762</td>\n",
       "      <td>Risankizumab</td>\n",
       "      <td>[inhibitor]</td>\n",
       "      <td>Peptides</td>\n",
       "      <td>drug</td>\n",
       "      <td>5970</td>\n",
       "      <td>Interleukin-23</td>\n",
       "      <td>target</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13423</th>\n",
       "      <td>DB15250</td>\n",
       "      <td>Vercirnon</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>drug</td>\n",
       "      <td>1610</td>\n",
       "      <td>C-C chemokine receptor type 9</td>\n",
       "      <td>target</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "       drug_ID              drug_name  drug_action drug_type snode_type  \\\n",
       "13414  DB14738  Turoctocog alfa pegol  [activator]  Peptides       drug   \n",
       "13415  DB14738  Turoctocog alfa pegol  [activator]  Peptides       drug   \n",
       "13416  DB14738  Turoctocog alfa pegol     [binder]  Peptides       drug   \n",
       "13417  DB14751   Mecasermin rinfabate    [agonist]  Peptides       drug   \n",
       "13418  DB14751   Mecasermin rinfabate               Peptides       drug   \n",
       "13419  DB14751   Mecasermin rinfabate               Peptides       drug   \n",
       "13420  DB14754           Solriamfetol                              drug   \n",
       "13421  DB14754           Solriamfetol                              drug   \n",
       "13422  DB14762           Risankizumab  [inhibitor]  Peptides       drug   \n",
       "13423  DB15250              Vercirnon                              drug   \n",
       "\n",
       "      target_ID                                      target_name tnode_type  \n",
       "13414      3551                            Coagulation factor IX     target  \n",
       "13415      3528                             Coagulation factor X     target  \n",
       "13416      3535                                      Prothrombin     target  \n",
       "13417      5465            Insulin-like growth factor 1 receptor     target  \n",
       "13418      6091                                 Insulin receptor     target  \n",
       "13419      5467  Cation-independent mannose-6-phosphate receptor     target  \n",
       "13420     11049            Sodium-dependent dopamine transporter     target  \n",
       "13421     11048       Sodium-dependent noradrenaline transporter     target  \n",
       "13422      5970                                   Interleukin-23     target  \n",
       "13423      1610                    C-C chemokine receptor type 9     target  "
      ]
     },
     "execution_count": 12,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "DB_df.tail(10)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 58,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Create DrugBank drug graph object\n",
    "DBdrug_target_graph = nx.from_pandas_edgelist(DB_df, source = 'drug_ID', target = 'target_ID', \n",
    "                                             edge_attr = ['drug_action'])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 59,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Node attributes: key is node, value is dict of attribute-value pairs\n",
    "snode_attr = {node: {'node_type': DB_df.iloc[i]['snode_type'], \n",
    "                     'drug_name': DB_df.iloc[i]['drug_name'],\n",
    "                     'drug_type': DB_df.iloc[i]['drug_type']} for i, node in enumerate(DB_df['drug_ID'])}\n",
    "tnode_attr = {node: {'node_type': DB_df.iloc[i]['tnode_type'], \n",
    "                     'gene_name': DB_df.iloc[i]['target_name']} for i, node in enumerate(DB_df['target_ID'])}\n",
    "\n",
    "nx.set_node_attributes(DBdrug_target_graph, snode_attr)\n",
    "nx.set_node_attributes(DBdrug_target_graph, tnode_attr)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 60,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Create edgelist file for DrugBank drug-targets\n",
    "filepath = '../Graphs/DrugBank_drug_target.edgelist'\n",
    "with open(filepath, 'wb') as file:\n",
    "    nx.write_edgelist(DBdrug_target_graph, file)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 61,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Combine DB drug-gene graph with OrphaNet disease-gene graph by gene nodes?\n",
    "ORDO_DB_graph = nx.compose(disease_gene_graph, DBdrug_target_graph)\n",
    "\n",
    "# Write to file\n",
    "filepath = '../Graphs/ORDO_DB_combined.edgelist'\n",
    "with open(filepath, 'wb') as file:\n",
    "    nx.write_edgelist(ORDO_DB_graph, file)\n",
    "    "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 27,
   "metadata": {},
   "outputs": [],
   "source": [
    "\"\"\"\n",
    "CHEMICAL-gene target from CTD\n",
    "***Gene IDs are NCBI identifiers, which are not in OrphaNet data...have to map if want to connect the graphs\n",
    "\n",
    "\"\"\"\n",
    "### THIS ALSO TAKES A WHILE TO RUN (not as long as DrugBank though)\n",
    "filepath = '../Data/CTD_chem_gene_ixns.csv'\n",
    "with open(filepath, 'r') as file:\n",
    "    reader = csv.reader(file, delimiter=',', quotechar='\"')\n",
    "    data_rows = [line for line in reader if line[0][0] != '#']\n",
    "    \n",
    "file.close()\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 45,
   "metadata": {
    "collapsed": true
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "835389\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>chem_name</th>\n",
       "      <th>chem_ID</th>\n",
       "      <th>CAS</th>\n",
       "      <th>gene</th>\n",
       "      <th>gene_ID</th>\n",
       "      <th>gene_forms</th>\n",
       "      <th>organism</th>\n",
       "      <th>organism_ID</th>\n",
       "      <th>ixn</th>\n",
       "      <th>ixn_actions</th>\n",
       "      <th>PMIDs</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>10074-G5</td>\n",
       "      <td>C534883</td>\n",
       "      <td></td>\n",
       "      <td>MYC</td>\n",
       "      <td>4609</td>\n",
       "      <td>protein</td>\n",
       "      <td>Homo sapiens</td>\n",
       "      <td>9606</td>\n",
       "      <td>10074-G5 analog results in decreased expressio...</td>\n",
       "      <td>decreases^expression</td>\n",
       "      <td>26036281</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>10074-G5</td>\n",
       "      <td>C534883</td>\n",
       "      <td></td>\n",
       "      <td>MYC</td>\n",
       "      <td>4609</td>\n",
       "      <td>protein</td>\n",
       "      <td>Homo sapiens</td>\n",
       "      <td>9606</td>\n",
       "      <td>10074-G5 results in decreased activity of MYC ...</td>\n",
       "      <td>decreases^activity</td>\n",
       "      <td>25716159</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>10074-G5</td>\n",
       "      <td>C534883</td>\n",
       "      <td></td>\n",
       "      <td>MYC</td>\n",
       "      <td>4609</td>\n",
       "      <td>protein</td>\n",
       "      <td>Homo sapiens</td>\n",
       "      <td>9606</td>\n",
       "      <td>10074-G5 results in decreased expression of MY...</td>\n",
       "      <td>decreases^expression</td>\n",
       "      <td>26036281</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12</th>\n",
       "      <td>10,11-dihydro-10,11-dihydroxy-5H-dibenzazepine...</td>\n",
       "      <td>C004822</td>\n",
       "      <td>35079-97-1</td>\n",
       "      <td>EPHX1</td>\n",
       "      <td>2052</td>\n",
       "      <td>gene</td>\n",
       "      <td>Homo sapiens</td>\n",
       "      <td>9606</td>\n",
       "      <td>[EPHX1 gene SNP affects the metabolism of carb...</td>\n",
       "      <td>affects^chemical synthesis|affects^metabolic p...</td>\n",
       "      <td>15692831</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13</th>\n",
       "      <td>10,11-dihydro-10,11-dihydroxy-5H-dibenzazepine...</td>\n",
       "      <td>C004822</td>\n",
       "      <td>35079-97-1</td>\n",
       "      <td>EPHX1</td>\n",
       "      <td>2052</td>\n",
       "      <td>protein</td>\n",
       "      <td>Homo sapiens</td>\n",
       "      <td>9606</td>\n",
       "      <td>[EPHX1 protein results in increased metabolism...</td>\n",
       "      <td>increases^chemical synthesis|increases^metabol...</td>\n",
       "      <td>15692831</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>14</th>\n",
       "      <td>10,11-dihydro-10-hydroxycarbamazepine</td>\n",
       "      <td>C039775</td>\n",
       "      <td></td>\n",
       "      <td>ABCB1</td>\n",
       "      <td>5243</td>\n",
       "      <td>protein</td>\n",
       "      <td>Homo sapiens</td>\n",
       "      <td>9606</td>\n",
       "      <td>ABCB1 protein results in increased transport o...</td>\n",
       "      <td>increases^transport</td>\n",
       "      <td>16190932</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>19</th>\n",
       "      <td>10-(2-pyrazolylethoxy)camptothecin</td>\n",
       "      <td>C534422</td>\n",
       "      <td></td>\n",
       "      <td>CASP3</td>\n",
       "      <td>836</td>\n",
       "      <td>protein</td>\n",
       "      <td>Homo sapiens</td>\n",
       "      <td>9606</td>\n",
       "      <td>10-(2-pyrazolylethoxy)camptothecin results in ...</td>\n",
       "      <td>increases^activity</td>\n",
       "      <td>18708040</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>20</th>\n",
       "      <td>10-(2-pyrazolylethoxy)camptothecin</td>\n",
       "      <td>C534422</td>\n",
       "      <td></td>\n",
       "      <td>CASP8</td>\n",
       "      <td>841</td>\n",
       "      <td>protein</td>\n",
       "      <td>Homo sapiens</td>\n",
       "      <td>9606</td>\n",
       "      <td>10-(2-pyrazolylethoxy)camptothecin results in ...</td>\n",
       "      <td>increases^activity</td>\n",
       "      <td>18708040</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>21</th>\n",
       "      <td>103D5R</td>\n",
       "      <td>C496879</td>\n",
       "      <td></td>\n",
       "      <td>AKT1</td>\n",
       "      <td>207</td>\n",
       "      <td>protein</td>\n",
       "      <td>Homo sapiens</td>\n",
       "      <td>9606</td>\n",
       "      <td>103D5R results in decreased phosphorylation of...</td>\n",
       "      <td>decreases^phosphorylation</td>\n",
       "      <td>15695405</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>22</th>\n",
       "      <td>103D5R</td>\n",
       "      <td>C496879</td>\n",
       "      <td></td>\n",
       "      <td>HIF1A</td>\n",
       "      <td>3091</td>\n",
       "      <td>protein</td>\n",
       "      <td>Homo sapiens</td>\n",
       "      <td>9606</td>\n",
       "      <td>103D5R inhibits the reaction [cobaltous chlori...</td>\n",
       "      <td>decreases^reaction|increases^expression</td>\n",
       "      <td>15695405</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                                            chem_name  chem_ID         CAS  \\\n",
       "2                                            10074-G5  C534883               \n",
       "3                                            10074-G5  C534883               \n",
       "4                                            10074-G5  C534883               \n",
       "12  10,11-dihydro-10,11-dihydroxy-5H-dibenzazepine...  C004822  35079-97-1   \n",
       "13  10,11-dihydro-10,11-dihydroxy-5H-dibenzazepine...  C004822  35079-97-1   \n",
       "14              10,11-dihydro-10-hydroxycarbamazepine  C039775               \n",
       "19                 10-(2-pyrazolylethoxy)camptothecin  C534422               \n",
       "20                 10-(2-pyrazolylethoxy)camptothecin  C534422               \n",
       "21                                             103D5R  C496879               \n",
       "22                                             103D5R  C496879               \n",
       "\n",
       "     gene gene_ID gene_forms      organism organism_ID  \\\n",
       "2     MYC    4609    protein  Homo sapiens        9606   \n",
       "3     MYC    4609    protein  Homo sapiens        9606   \n",
       "4     MYC    4609    protein  Homo sapiens        9606   \n",
       "12  EPHX1    2052       gene  Homo sapiens        9606   \n",
       "13  EPHX1    2052    protein  Homo sapiens        9606   \n",
       "14  ABCB1    5243    protein  Homo sapiens        9606   \n",
       "19  CASP3     836    protein  Homo sapiens        9606   \n",
       "20  CASP8     841    protein  Homo sapiens        9606   \n",
       "21   AKT1     207    protein  Homo sapiens        9606   \n",
       "22  HIF1A    3091    protein  Homo sapiens        9606   \n",
       "\n",
       "                                                  ixn  \\\n",
       "2   10074-G5 analog results in decreased expressio...   \n",
       "3   10074-G5 results in decreased activity of MYC ...   \n",
       "4   10074-G5 results in decreased expression of MY...   \n",
       "12  [EPHX1 gene SNP affects the metabolism of carb...   \n",
       "13  [EPHX1 protein results in increased metabolism...   \n",
       "14  ABCB1 protein results in increased transport o...   \n",
       "19  10-(2-pyrazolylethoxy)camptothecin results in ...   \n",
       "20  10-(2-pyrazolylethoxy)camptothecin results in ...   \n",
       "21  103D5R results in decreased phosphorylation of...   \n",
       "22  103D5R inhibits the reaction [cobaltous chlori...   \n",
       "\n",
       "                                          ixn_actions     PMIDs  \n",
       "2                                decreases^expression  26036281  \n",
       "3                                  decreases^activity  25716159  \n",
       "4                                decreases^expression  26036281  \n",
       "12  affects^chemical synthesis|affects^metabolic p...  15692831  \n",
       "13  increases^chemical synthesis|increases^metabol...  15692831  \n",
       "14                                increases^transport  16190932  \n",
       "19                                 increases^activity  18708040  \n",
       "20                                 increases^activity  18708040  \n",
       "21                          decreases^phosphorylation  15695405  \n",
       "22            decreases^reaction|increases^expression  15695405  "
      ]
     },
     "execution_count": 45,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Create CTD data frame\n",
    "CTD_cols = ['chem_name', 'chem_ID', 'CAS', 'gene', 'gene_ID', 'gene_forms', \n",
    "           'organism', 'organism_ID', 'ixn', 'ixn_actions', 'PMIDs']\n",
    "CTD_df = pd.DataFrame(data_rows, columns = CTD_cols)\n",
    "\n",
    "# Keep only interactions with human genes ('organism_ID' == '9606')\n",
    "CTD_df = CTD_df[CTD_df['organism_ID'] == '9606']\n",
    "print(len(CTD_df))\n",
    "CTD_df.head(10)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Create graph from CTD dataframe\n",
    "chem_gene_graph = nx.from_pandas_edgelist(CTD_df, source = 'chem_ID', target = 'gene_ID', \n",
    "                                             edge_attr = ['assoc_type', 'assoc_status'])\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 82,
   "metadata": {},
   "outputs": [],
   "source": [
    "\"\"\"\n",
    "DRUG-gene target from KEGG*\n",
    "*Combine with DrugBank??\n",
    "Honestly KEGG's format is a pain though\n",
    "########################\n",
    "\n",
    "filepath = \"../Data/KEGG_drug\"\n",
    "with open(filepath, 'r') as file:\n",
    "    # Separate entries (///), then create dictionary\n",
    "    entries = file.read().split('\\n///\\n')\n",
    "    entrylist = list()\n",
    "    for i, entry in enumerate(entries):\n",
    "        entry = entry.split('\\n')\n",
    "        entrydict = dict()\n",
    "        for row in entry:\n",
    "            pattern = '\\s{3,}'\n",
    "            row = re.split(pattern, row)\n",
    "            if row[0] != '':\n",
    "                key = row[0]\n",
    "                entrydict[key] = row[1:]\n",
    "            else:\n",
    "                if key in entrydict.keys():\n",
    "                    entrydict[key].append(row[1:])  \n",
    "        entrylist.append(entrydict)\n",
    "\n",
    "file.close()\n",
    "\n",
    "# Only keep drug entries that contain target genes\n",
    "entrylist = [entry for entry in entrylist if 'TARGET' in list(entry.keys())]\n",
    "\"\"\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 80,
   "metadata": {
    "collapsed": true
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "{'ENTRY': ['D00045', 'Drug'],\n",
       " 'NAME': ['Adenosine (JAN/USP);', ['Adenocard (TN);'], ['Adenoscan (TN)']],\n",
       " 'FORMULA': ['C10H13N5O4'],\n",
       " 'EXACT_MASS  267.0968': [],\n",
       " 'MOL_WEIGHT  267.2413': [],\n",
       " 'REMARK': ['Same as: C00212',\n",
       "  ['Therapeutic category: 7990'],\n",
       "  ['ATC code: C01EB10'],\n",
       "  ['Chemical structure group: DG00243'],\n",
       "  ['Product (DG00243): D00045<JP/US> D02300<JP>']],\n",
       " 'EFFICACY': ['Antiarrhythmic, Cardiac depressant, Diagnostic aid (cardiac stress test), Adenosine receptor agonist'],\n",
       " 'TARGET': ['ADORA [HSA:134 135 136 140] [KO:K04265 K04266 K04267 K04268]'],\n",
       " 'INTERACTION  ': [],\n",
       " 'DBLINKS': ['CAS: 58-61-7',\n",
       "  ['PubChem: 7847113'],\n",
       "  ['ChEBI: 16335'],\n",
       "  ['ChEMBL: CHEMBL477'],\n",
       "  ['DrugBank: DB00640'],\n",
       "  ['PDB-CCD: ADN'],\n",
       "  ['LigandBox: D00045'],\n",
       "  ['NIKKAJI: J4.501B']],\n",
       " 'ATOM': ['19',\n",
       "  ['1', 'N4y N', '27.7735  -15.1762'],\n",
       "  ['2', 'C8y C', '28.7550  -16.0205'],\n",
       "  ['3', 'C1y C', '26.4808  -15.5916'],\n",
       "  ['4', 'C8x C', '28.3947  -13.8350'],\n",
       "  ['5', 'C8y C', '30.0785  -15.2503'],\n",
       "  ['6', 'N5x N', '28.7377  -17.4480'],\n",
       "  ['7', 'O2x O', '25.3403  -14.7435'],\n",
       "  ['8', 'C1y C', '26.0539  -16.8432'],\n",
       "  ['9', 'N5x N', '29.8363  -13.9950'],\n",
       "  ['10  C8y C', '31.2648  -15.9301'],\n",
       "  ['11  C8x C', '29.9894  -18.1759'],\n",
       "  ['12  C1y C', '24.2407  -15.5331'],\n",
       "  ['13  C1y C', '24.6677  -16.8432'],\n",
       "  ['14  O1a O', '26.4632  -18.1767'],\n",
       "  ['15  N5x N', '31.2695  -17.4246'],\n",
       "  ['16  N1a N', '32.4508  -15.2334'],\n",
       "  ['17  C1b C', '22.9364  -15.1236'],\n",
       "  ['18  O1a O', '24.2231  -18.1709'],\n",
       "  ['19  O1a O', '21.8253  -15.9600']],\n",
       " 'BOND': ['21',\n",
       "  ['1', '1', '2 1'],\n",
       "  ['2', '3', '1 1 #Up'],\n",
       "  ['3', '1', '4 1'],\n",
       "  ['4', '2', '5 2'],\n",
       "  ['5', '2', '6 1'],\n",
       "  ['6', '3', '7 1'],\n",
       "  ['7', '3', '8 1'],\n",
       "  ['8', '4', '9 2'],\n",
       "  ['9', '5  10 1'],\n",
       "  ['10', '6  11 2'],\n",
       "  ['11', '7  12 1'],\n",
       "  ['12', '8  13 1'],\n",
       "  ['13', '8  14 1 #Down'],\n",
       "  ['14', '10  15 2'],\n",
       "  ['15', '10  16 1'],\n",
       "  ['16', '12  17 1 #Up'],\n",
       "  ['17', '13  18 1 #Down'],\n",
       "  ['18', '17  19 1'],\n",
       "  ['19', '5', '9 1'],\n",
       "  ['20', '11  15 1'],\n",
       "  ['21', '12  13 1']]}"
      ]
     },
     "execution_count": 80,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# KEGG API for retrieving target gene references\n",
    "# URL = \"http://rest.kegg.jp/find/genes/{}\"\n",
    "# entryID = ''"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "\"\"\"\n",
    "Gene pathways from KEGG\n",
    "\n",
    "\"\"\"\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "\"\"\"\n",
    "Drug chemical/molecular structure data (ChEMBL?)\n",
    "\n",
    "\"\"\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "\"\"\"\n",
    "REFERENCES:\n",
    "\n",
    "Orphanet data: http://www.orphadata.org/cgi-bin/index.php\n",
    "DrugBank data (large, like 1.5 GB) + XSD file: https://www.drugbank.ca/releases/latest\n",
    "CTD data: http://ctdbase.org/downloads/\n",
    "KEGG DRUG data file: https://www.genome.jp/kegg/drug/\n",
    "KEGG PATHWAYS: \n",
    "\n",
    "Reading from Orphanet disease-gene XML file\n",
    ">https://stackoverflow.com/questions/2148119/how-to-convert-an-xml-string-to-a-dictionary\n",
    "    >http://code.activestate.com/recipes/410469-xml-as-dictionary/\n",
    "    \n",
    "\"\"\""
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.7.1"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
